File: test-suite-python/src/test/python/micronaut/docs/config/properties/EngineConfig.py
from typing import Annotated

# tag::imports[]
from micronaut.context.annotation import ConfigurationProperties

from micronaut.context.annotation import Requires
from jakarta.validation.constraints import Min, NotBlank
# end::imports[]


@Requires(property="spec.name", value="VehiclePropertiesSpec")
# tag::class[]
@ConfigurationProperties("my.engine")  # <1>
class EngineConfig:
    @ConfigurationProperties("crank-shaft")
    class CrankShaft:  # <4>
        rod_length: float | None = None  # <5>

    manufacturer: Annotated[str, NotBlank] = "Ford"  # <2> <3>
    cylinders: Annotated[int, Min(1)]
    crank_shaft: CrankShaft = CrankShaft()
# end::class[]


CrankShaft = EngineConfig.CrankShaft

File: test-suite-scala/src/test/scala/io/micronaut/docs/config/immutable/EngineConfig.scala
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.docs.config.immutable

// tag::imports[]
import io.micronaut.context.annotation.ConfigurationProperties
import io.micronaut.context.annotation.Requires
import io.micronaut.core.bind.annotation.Bindable
import org.jspecify.annotations.Nullable

import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import java.util.Optional
// end::imports[]

@Requires(property = "spec.name", value = "VehicleImmutableSpec")
// tag::class[]
@ConfigurationProperties("my.engine") // <1>
case class EngineConfig( // <2>
    @Bindable(defaultValue = "Ford") @NotBlank manufacturer: String, // <3>
    @Min(value = 1L) cylinders: Int, // <4>
    @NotNull crankShaft: EngineConfig.CrankShaft // <5>
)

object EngineConfig:
  @ConfigurationProperties("crank-shaft")
  case class CrankShaft(
      @Nullable rodLength: java.lang.Double // <6>
  ):
    def getRodLength: Optional[java.lang.Double] =
      Optional.ofNullable(rodLength)
// end::class[]
